casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala

```diff
@@ -91,32 +127,46 @@ object EstimatorHelper {
-
-      producesInCommEvents = commEvents.flatMap(_.produces)
-      consumeInCommEvents  = commEvents.map(_.consume)
+      allProduceEvents = ancestorEvents.collect { case p: Produce => p }
+      allConsumeEvents = ancestorEvents.collect { case c: Consume => c }
+      allCommEvents    = ancestorEvents.collect { case c: COMM    => c }
+      (volatileCommEvents, nonVolatileCommEvents) = allCommEvents
+        .partition(isVolatile(_, allConsumeEvents, allProduceEvents))
+      producesInVolatileCommEvents = volatileCommEvents.flatMap(_.produces)
+      consumesInVolatileCommEvents = volatileCommEvents.map(_.consume)
+      produceEvents                = allProduceEvents.filterNot(producesInVolatileCommEvents.contains(_))
+      consumeEvents                = allConsumeEvents.filterNot(consumesInVolatileCommEvents.contains(_))
+    } yield (produceEvents, consumeEvents, nonVolatileCommEvents)
 
-      ancestorProduceChannels = produceEvents
-        .filterNot(producesInCommEvents.contains(_))
-        .map(_.channelsHash)
-        .toSet
+  private[this] def extractJoinedChannels(b: BlockEvents): Set[Blake2b256Hash] = {
+    def joinedChannels(consumes: Set[Consume]) =
+      consumes.withFilter(Consume.hasJoins).flatMap(_.channelsHashes)
+    b match {
+      case (_, consumes, comms) =>
+        joinedChannels(consumes) ++ joinedChannels(comms.map(_.consume))
+    }
+  }
 
-      ancestorConsumeChannels = consumeEvents
-        .filterNot(consumeInCommEvents.contains(_))
-        .flatMap(_.channelsHashes)
-        .toSet
+  private[this] def tuplespaceEventsPerChannel(
+      b: BlockEvents
+  ): Map[Blake2b256Hash, Set[TuplespaceEvent]] =
+    b match {
+      case (produces, consumes, comms) =>
+        val nonPersistentProducesInComms = comms.flatMap(_.produces).filterNot(_.persistent)
+        val nonPersistentConsumesInComms = comms.map(_.consume).filterNot(_.persistent)
+        val freeProduces                 = produces.filterNot(nonPersistentProducesInComms.contains)
+        val freeConsumes                 = consumes.filterNot(nonPersistentConsumesInComms.contains)
 
-      ancestorCommChannels = commEvents
-        .filterNot(isVolatile(_, consumeEvents, produceEvents))
-        .flatMap { comm =>
-          comm.consume.channelsHashes ++ comm.produces.map(_.channelsHash)
-        }
-        .toSet
+        val produceEvents = freeProduces.map(TuplespaceEvent.from(_))
+        val consumeEvents = freeConsumes.flatMap(TuplespaceEvent.from(_))
+        val commEvents    = comms.flatMap(TuplespaceEvent.from(_, produces))
 
-      ancestorChannels = ancestorProduceChannels ++ ancestorConsumeChannels ++ ancestorCommChannels
-      res              = ancestorChannels.map(_.toByteString)
-    } yield res
+        (produceEvents
+          .combine(consumeEvents)
+          .combine(commEvents))
+          .groupBy(_._1)
+          .mapValues[Set[TuplespaceEvent]](_.map(_._2))
```

Comment:
```
        (produceEvents ++ consumeEvents ++ commEvents)
          .groupBy(_._1)
          .mapValues[Set[TuplespaceEvent]](_.map(_._2))
```

if you insist on using cats, please use `|+|`

Reply:
`combine` comes from before refactoring. `++` will work fine do now